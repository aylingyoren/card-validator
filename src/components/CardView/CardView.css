.card-body {
  position: relative;
  margin-bottom: 50px;
  border-radius: 15px;
  border: 3px solid rgb(50, 50, 50);
  background: linear-gradient(to bottom right, darkcyan, darkblue);
  width: 500px;
  height: 300px;
}

.light-card-border {
  border: 2px solid rgb(250, 250, 250);
}

.card-number,
.card-holder,
.exp-date,
.visa-img,
.chip-img,
.cvv {
  position: absolute;
}

.card-number,
.card-holder,
.exp-date,
.cvv {
  color: rgb(250, 250, 250);
}

.card-number {
  margin: 0;
  top: 50%;
  left: 15%;
  -webkit-transform: translateY(-50%);
  width: 100%;
  font-size: 30px;
}

.visa-img {
  top: -10px;
  right: 10px;
  width: 130px;
}

.chip-img {
  top: 10px;
}

.card-holder,
.exp-date,
.cvv {
  bottom: 30px;
  font-size: 18px;
}
.card-holder,
.chip-img {
  left: 30px;
}

.exp-date {
  left: 250px;
}

.cvv {
  right: 30px;
}
