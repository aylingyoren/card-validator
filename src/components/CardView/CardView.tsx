import React from "react";
import { MyFormValues } from "../Card/Card";
import "./CardView.css";

export interface CardViewProps {
  values: MyFormValues;
  isDarkModeActive: boolean;
}

function CardView({ values, isDarkModeActive }: CardViewProps) {
  const { cardNumber, cardHolder, month, year, CVV } = values;
  return (
    <div
      className={isDarkModeActive ? "card-body light-card-border" : "card-body"}
      data-testid="card-body"
    >
      <h2 className="card-number" data-testid="card-number">
        {cardNumber ? cardNumber : "0000 0000 0000 0000"}
      </h2>
      {/* <h2 className="card-number" data-testid="card-number">{cardNumber? `${cardNumber.slice(0,4)} ${cardNumber.slice(4,8)} ${cardNumber.slice(8,12)} ${cardNumber.slice(12,16)}`  : '0000 0000 0000 0000'}</h2> */}
      <div className="card-holder" data-testid="card-holder">
        {cardHolder ? cardHolder.toUpperCase() : "NAME SURNAME"}
      </div>
      <div className="exp-date" data-testid="exp-date">
        {month ? month : "00"} / {year ? year : "00"}
      </div>
      <div className="cvv" data-testid="cvv">
        CVV: {CVV ? CVV : "000(0)"}
      </div>
      <img
        src={require("../../img/chip-card.png")}
        alt="card chip"
        className="chip-img"
        data-testid="chip-img"
      />
      <img
        src={require("../../img/visa.png")}
        alt="visa logo"
        className="visa-img"
        data-testid="visa-img"
      />
    </div>
  );
}

export default CardView;
