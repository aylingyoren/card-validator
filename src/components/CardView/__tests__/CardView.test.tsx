import React from "react";
import renderer from "react-test-renderer";
import {
  render,
  screen,
  getByText,
  fireEvent,
  waitFor,
} from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import { configure, shallow } from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import CardView, { CardViewProps } from "../CardView";

// configure({ adapter: new Adapter() });

// describe("CardView component", () => {
//   const props: CardViewProps = {
//     values: {
//       cardNumber: "1234567812345678",
//       cardHolder: "Tet Test",
//       month: "02",
//       year: "26",
//       CVV: "1234",
//     },
//     isDarkModeActive: false,
//   };

//   const component = shallow(<CardView {...props} />);

//   it("renders the <CardView /> component", () => {
//     expect(component.exists("card-body")).toBe(true); // false
//   });

//   // it('renders user name when props passed', () => {
//   //   const link = component.find('.header__login');
//   //   expect(link.text()).toBe(props.user.name);
//   // });
// });

// __________

const values = {
  cardNumber: "1234567812345678",
  cardHolder: "Alec Bow",
  month: "02",
  year: "23",
  CVV: "123",
};
const isDarkModeActive = false;

describe("CardView tests", () => {
  it("displays card binding", async () => {
    render(<CardView values={values} isDarkModeActive={isDarkModeActive} />);
    await waitFor(() => {
      expect(screen.getByText(values.cardNumber)).toBeInTheDocument();
    });
  });

  it("displays card holder in upper case", async () => {
    render(<CardView values={values} isDarkModeActive={isDarkModeActive} />);
    await waitFor(() => {
      expect(
        screen.getByText(values.cardHolder.toUpperCase())
      ).toBeInTheDocument();
    });
  });
});

// __________

function renderCardView(props: Partial<CardViewProps> = {}) {
  const defaultProps: CardViewProps = {
    values: {
      cardNumber: "",
      cardHolder: "",
      month: "",
      year: "",
      CVV: "",
    },
    isDarkModeActive: false,
  };
  return render(<CardView {...defaultProps} {...props} />);
}

test("should display a blank card form, with dark theme set by default", async () => {
  const { findByTestId } = renderCardView();

  const cardBody = await findByTestId("card-body");
  const cardNumber = await findByTestId("card-number");
  const visaImg = await findByTestId("visa-img");

  expect(cardBody).toBeVisible();
  expect(cardBody).toContainElement(cardNumber);
  expect(cardNumber).toContainHTML("0000 0000 0000 0000");
  expect(cardNumber).not.toContainHTML("0000000000000000");
  expect(visaImg).toHaveAccessibleName("visa logo");
});

it("renders correctly when there are no items", () => {
  const defaultProps: CardViewProps = {
    values: {
      cardNumber: "",
      cardHolder: "",
      month: "",
      year: "",
      CVV: "",
    },
    isDarkModeActive: false,
  };

  const tree = renderer.create(<CardView {...defaultProps} />).toJSON();
  expect(tree).toMatchSnapshot();
});

// test("CardView changes the class when hovered", () => {
//   const component = renderer.create(<CardView values={} isDarkModeActive />);
//   let tree = component.toJSON();
//   expect(tree).toMatchSnapshot();

//   // manually trigger the callback
//   //   tree.props.something();
//   // re-rendering
//   tree = component.toJSON();
//   expect(tree).toMatchSnapshot();

//   // manually trigger the callback
//   //   tree.props.somethingElse();
//   // re-rendering
//   tree = component.toJSON();
//   expect(tree).toMatchSnapshot();
// });
