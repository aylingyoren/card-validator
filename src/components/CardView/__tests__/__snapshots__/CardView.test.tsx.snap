// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when there are no items 1`] = `
<div
  className="card-body"
  data-testid="card-body"
>
  <h2
    className="card-number"
    data-testid="card-number"
  >
    0000 0000 0000 0000
  </h2>
  <div
    className="card-holder"
    data-testid="card-holder"
  >
    NAME SURNAME
  </div>
  <div
    className="exp-date"
    data-testid="exp-date"
  >
    00
     / 
    00
  </div>
  <div
    className=""
    data-testid="cvv"
  >
    CVV: 
    000(0)
  </div>
  <img
    alt="card chip"
    className=""
    data-testid="chip-img"
    src="chip-card.png"
  />
  <img
    alt="visa logo"
    className=""
    data-testid="visa-img"
    src="visa.png"
  />
</div>
`;
